src/coreclr/jit/emitxarch.cpp

```diff
@@ -12598,9 +12694,13 @@ size_t emitter::emitOutputInstr(insGroup* ig, instrDesc* id, BYTE** dp)
             // the loop alignment pseudo instruction
             if (ins == INS_align)
             {
-                sz  = SMALL_IDSC_SIZE;
-                dst = emitOutputNOP(dst, (-(int)(size_t)dst) & 0x0f);
-                assert(((size_t)dst & 0x0f) == 0);
+                sz = sizeof(instrDescAlign);
+                // IG can be marked as not needing alignment after emitting align instruction
+                // In such case, skip outputting alignment.
```

Comment:
Is this a "todo"?  We won't ever skip.

Reply:
In `emitLoopAlignAdjustments()`, if we detect that there is no padding needed, we set the `idCodeSize` to 0 and remove the `IGF_LOOP_ALIGN` flag. It is just to avoid going inside `emitOutputAlign()` for such cases.

Reply:
I just recalled that another reason that I do this is because I started optimizing the marking of `emitLastAlignedIgNum` only the one that has non-zero padding. Earlier, I was marking every IG for which we add `align` instruction. But ideally, it should be the last IG that added a non-zero padding. With the 0b85575acad change, we see some good improvements in code size because we won't have to compensate the over-estimation for `align` that will add zero padding. I had this earlier, but backed out for some reason. I am adding it back. With this change, we need to make sure that we do not enter `emitOutputAlign()` because if we, the `paddingNeeded` will be non-zero since we stopped compensating the overestimation.

```
Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 51641301
Total bytes of diff: 51636950
Total bytes of delta: -4351 (-0.01% of base)
    diff is an improvement.

Top file improvements (bytes):
        -727 : System.Private.CoreLib.dasm (-0.01% of base)
        -700 : System.Data.Common.dasm (-0.05% of base)
        -609 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm (-0.02% of base)
        -276 : FSharp.Core.dasm (-0.01% of base)
        -238 : Microsoft.CodeAnalysis.CSharp.dasm (-0.01% of base)
        -217 : System.Private.Xml.dasm (-0.01% of base)
        -147 : System.Memory.dasm (-0.04% of base)
        -110 : System.Linq.dasm (-0.01% of base)
         -97 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.00% of base)
         -90 : System.Threading.Tasks.Dataflow.dasm (-0.01% of base)
         -77 : Microsoft.VisualBasic.Core.dasm (-0.02% of base)
         -75 : System.Drawing.Common.dasm (-0.02% of base)
         -74 : Microsoft.CodeAnalysis.dasm (-0.00% of base)
         -72 : System.Runtime.Numerics.dasm (-0.10% of base)
         -57 : System.Formats.Asn1.dasm (-0.05% of base)
         -57 : System.Collections.dasm (-0.01% of base)
         -52 : System.Private.DataContractSerialization.dasm (-0.01% of base)
         -43 : System.Text.RegularExpressions.dasm (-0.02% of base)
         -41 : System.Collections.Immutable.dasm (-0.00% of base)
         -39 : System.Private.Uri.dasm (-0.04% of base)
```

Reply:
Just to confirm, I measured the overestimate compensation we add before and after this particular change. Before this change, we added `2189 bytes` of compensation, but after this change, we just ended up adding `619 bytes` of compensation.